docs/proposals/202003_thanos_rules_federation: initial commit (#2263)
thanos-io/thanos: docs/proposals/202003_thanos_rules_federation.md

```proto
service Rules {
  /// Rules has info for all rules.
  rpc Rules(RulesRequest) returns (stream RulesResponse);
}
```

Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based in its local state.

Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:

* Deduplication happens first the rule group level. The identifier is the group name.
* Then, per group name deduplication happens on the rule level, where:

1. the rule type (recording rule vs. alerting rule)
2. the rule name
3. the rule label names

are being used as the deduplication identifier. Disjunct entries are simply merged by adding them to the result set.

Thanos query presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.

> Additionally Thanos Querier gains a new `--rule.replica-label` command line argument which falls back to `--query.replica-label` if unset.

Reviewer: Could we also describe how are we going to set the replica label in sidecar and thanos ruler?

Right now when we do Prometheus metric dedup, we set it in `external_labels` of prometheus config:
```
  external_labels:
    replica: replica-$(HOSTNAME)
```

Reviewer 2: This is too complex.. Why we cannot have just `query` replica and only that?

Reviewer 2: so we can reuse external labels for sidecar and thanos rule has `--label` already?

Reviewer: Yup, great idea. Let's add this to the proposal.

Author: ok, sounds good, will update the design doc :+1:

Author: Added a small section about Thanos Ruler `--label` and `external_labels` that will be reused. How about Thanos Querier though? As far as I see it does not have a `--label` command line argument, I guess we need one here too?

Scenario 3:

If, under the above conditions a rule is a candidate for deduplication, finally the rule `expr` and `for` fields are being considered for deduplication.

Given the following stream of incoming alerting rules will also result in two independent alerting rules as both the `expr` and `for` fields differ:
```text
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate1h) > (14.40 * 0.01000)
      and
      sum(apiserver_request:burnrate5m) > (14.40 * 0.01000)
    for: 2m
    labels:
      severity: critical
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate6h) > (6.00 * 0.01000)
      and
      sum(apiserver_request:burnrate30m) > (6.00 * 0.01000)
    for: 15m
    labels:
      severity: critical
```

### Type and Info

To accommodate for future additional APIs alongside with the core Rules proto changes an enhancement of the protobuf Info and Type structures are suggested.

>  The current `StoreType` enum is renamed to `Type`. This retains binary compatibility with older clients:

Reviewer: Why do we need this? Can this be a different prosposal? as right now the main proposal introduces new Service and doesn't change existing protos or services.

Author: yes, I agree, this came up during a pairing session with @bwplotka but I agree we should handle this in a separate proposal.

Reviewer: After rethinking this, if we merge this proposal we don't have a way to do discover Rule apis, so I'm fine with this being in this proposal. Was just a bit unexpected addition after doing a bunc reviews :) Sorry!

Reviewer: I guess my question is then what do we do with the `--store` , `--store.sd-files=` flags? As this need to renamed too now :)

Reviewer 2: Yea, we need to leave it I guess for now.

Ok, let's decide what do in this Proposal. I think we can limit this proposal for now to NO standalone RuleAPI for now.

Then we can move those proto info changes to Future Work so our ideas are maintained! :hugs: 

WDYT @s-urbaniak

Author: > Was just a bit unexpected addition after doing a bunch back and forth reviews :) Sorry!

Nothing to be sorry about! I am very grateful for the thorough review, time and the great catches :bow: :heart: !

Regarding `--store` , and `--store.sd-files=`. Hmm ... i'd say we can leave them as we are introducing an orthognal `--rule` flag, no? cc @bwplotka

Reviewer 2: I would keep it simple for now

Author: ok, leaving as is :+1:

 ```diff
-enum StoreType {
+enum Type {
   UNKNOWN = 0;
   QUERY = 1;
   RULE = 2;
```

 The current fiels in `InfoResponse` connected to Store APIs are deprecated and dedicated new API sub types are proposed:

 ```diff
 message InfoResponse {
   // Deprecated. Use label_sets instead.
   repeated Label labels = 1 [(gogoproto.nullable) = false];
+  // Deprecated. Will be removed in favor of StoreInfoResponse in the future.
   int64 min_time = 2;
+  // Deprecated. Will be removed in favor of StoreInfoResponse in the future.
   int64 max_time = 3;
-  StoreType storeType  = 4;
+  Type type  = 4;
   // label_sets is an unsorted list of `LabelSet`s.
+  // Deprecated. Will be removed in favor of StoreInfoResponse in the future.
   repeated LabelSet label_sets = 5 [(gogoproto.nullable) = false];
+
+  StoreInfoResponse store = 6;

> +  RulesInfoResponse rules = 7;

Reviewer: What is this? Why is this needed? Where are the protos for this?

Author: Agreed, see https://github.com/thanos-io/thanos/pull/2263#discussion_r410143232, let's handle this in a separate proposal.

Reviewer 2: Maybe let's put into # Future so it;s not lost???

Reviewer: +1 on future

Author: Moved to a `## Future` section at the end :+1:

---
title: Global / Federated Rules API
type: proposal
menu: proposals
status: accepted

> owner: s-urbaniak,bwplotka

Reviewer 2: ```suggestion
owner: s-urbaniak
```

I think essentially this will render wrongly on website (: We need one ;p

Reviewer 2: Happy to co-own this though :+1:

Author: Happy to take responsibility for these changes :+1:

The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alerting rules.

```proto
service Rules {
  /// Rules has info for all rules.
  rpc Rules(RulesRequest) returns (stream RulesResponse);
}
```

Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based in its local state.

Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:

* Deduplication happens first the rule group level. The identifier is the group name.
* Then, per group name deduplication happens on the rule level, where:

1. the rule type (recording rule vs. alerting rule)
2. the rule name
3. the rule label names
4. the rule expression `expr` field
5. the alerting rule `for` field

are being used as the deduplication identifier. Disjunct entries are simply merged by adding them to the result set.

> 

Reviewer 2: Can we mentioned that `Info` is shared with StoreAPI for now?

Author: fixed

```text
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate1h) > (14.40 * 0.01000)
      and
      sum(apiserver_request:burnrate5m) > (14.40 * 0.01000)
    for: 2m
    labels:
      severity: critical
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate6h) > (6.00 * 0.01000)
      and
      sum(apiserver_request:burnrate30m) > (6.00 * 0.01000)
    for: 15m
    labels:
      severity: critical
```

## Alternatives

* Cortex contains a sharded Ruler. Assigning rules to shards is done via Consul, though a gossip implementation is under development. Shards do not communicate with other shards. Rules come from a store (e.g. a Postgres database).

## Work Plan

* Implement a new flag `--rule` in Thanos Querier which registers RulesAPI endpoints.
* Implement a new flag `--rule.replica-label` in Thanos Querier.
* Implement RulesAPI backends in sidecar, query, rule.
* Feature branch: https://github.com/thanos-io/thanos/pull/2200

## Future

> 

Reviewer: Let's add sentence to clarify this:

These changes are suggestions which we will need to be discussed in future and are not part of the proposal implementation.

Author: fixed :+1:

---
title: Global / Federated Rules API
type: proposal
menu: proposals
status: accepted
owner: s-urbaniak
---

### Related Tickets

https://github.com/thanos-io/thanos/issues/1375

## Summary

Thanos allows a global query view for the Prometheus series. This means discovering, connecting various (often remote) "leafs" components and aggregating series data from them.

Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.

## Motivation

> Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and Thanos Ruler instances.

Reviewer 3: Not sure about this one. That sentence is weird.

```suggestion
Deployment scenarios with various levels of federated setups are not unusual. Some deployments leverage Thanos Ruler to query multiple data sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology, a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and Thanos Ruler instances.
```

Author: committed your suggestion :+1:

Pitfalls of current solutions:
* It’s tedious to manually visit each leaf and we are lazy.
* Visiting leaf instances (be it Thanos Ruler or dedicated Prometheus endpoints) have to be implemented manually.

## Goals

* Present other Prometheus/Thanos Resources in a global view.
    * This also means an up-to-date view (e.g statuses)
* Simple to run if you already run Thanos.
    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
    * Potentially use the same connections, discovered targets by Querier.

## Verification

* Unit tests which would fire up a dummy Rules API and check different scenarios.
* Ad-hoc testing.

## Proposal

### Rules API

> We propose to add following proto service and propagate all using existing components that has rules, so:

Author: thank you! fixed

* Sidecar
* Querier (for federation)
* Ruler

The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alerting rules.

```proto
service Rules {
  /// Rules has info for all rules.
  rpc Rules(RulesRequest) returns (stream RulesResponse);
}
```

> Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based in its local state.

Reviewer 3: ```suggestion
Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based on its local state.
```

Author: thank you, fixed :-)

> Thanos Querier fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:

Reviewer 2: There is missing decision item:

Can user specify --rule X without --store X? Is this planned to be implemented?

Author: @bwplotka i suggest to have completely independent settings as a future enhancement. For now, we initially discussed to have rule endpoints be a strict subset of store endpoints which makes the initial code handling much easier (piggy packing on the existing store health checks)

Author: Concrete implication for the initial implementation would be: if user does specify `--rule` which do not fit as a subset of the given `--store` (and other related store discover settings), we'd log out that fact. Does that sound reasonable/

Reviewer 2: Could be as a start yes, but it's not stated yet clear here. (: Can we clarify?

Author: Clarified :+1: :green_heart:

> Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:

Author: fixed

> * Deduplication happens first the rule group level. The identifier is the group name.

Author: fixed

These may be remote Thanos Ruler, Prometheus Sidecar, or even Thanos Querier instances.
The merging Thanos Querier has `--rule.replica-label=replica`

Scenario 1:

As specified, the rule type and then rule name is used for deduplication.

Given the following stream of incoming rule groups and containing recording/alerting rules:
```text
group: a
   recording:<name:"a1" last_evaluation:<seconds:1 > >
   alert:    <name:"r1" last_evaluation:<seconds:1 > >
group: b
   recording:<name:"r1" last_evaluation:<seconds:1 > >
group: a
   recording:<name:"r2" last_evaluation:<seconds:1 > >
```

The output becomes:
```text
group: a
   alert:    <name:"a1" last_evaluation:<seconds:1 > >
   recording:<name:"r1" last_evaluation:<seconds:1 > >
   recording:<name:"r2" last_evaluation:<seconds:1 > >
group: b
   recording:<name:"r1" last_evaluation:<seconds:1 > >
```

Note in the example above how the recording rule `r1` is not deduplicated as it is contained in two different groups.

Scenario 2:

The next level of deduplication is governed by the label/value set of the underlying recording/alerting rule while respecting the replica label.

> For a given conflict, the youngest is preferred. For alerting rules, the youngest firing rule is preferred.

Reviewer 3: You mean this, right, as in the example that immediately follows this?

```suggestion
For a given recording rules conflict, the oldest is preferred. For alerting rules, the youngest firing rule is preferred.
```

Author: 2 seconds is meant to be younger than 1 second :-) I meant youngest, yes. I'll change the examples to use full dates instead.

group: b
   recording:<name:"r1" last_evaluation:<seconds:1 > >
```

Note in the example above how the recording rule `r1` is not deduplicated as it is contained in two different groups.

Scenario 2:

The next level of deduplication is governed by the label/value set of the underlying recording/alerting rule while respecting the replica label.
For a given conflict, the youngest is preferred. For alerting rules, the youngest firing rule is preferred.

Given the following stream of incoming recording rules:
```text
group: a
   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<2006-01-02T10:00:00> >
group: a
   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >
```

The output becomes:
```text
group: a
   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >
```

Given the following stream of incoming alerting rules:
```text
group: a
   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<2006-01-02T10:00:00> >
group: a

>    alert:<             name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >

Reviewer 3: Where has the state disappeared? :smile: 

```suggestion
   alert:<state:PENDING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<2006-01-02T10:01:00> >
```

Author: sounds good, added, thank you!

---
title: Global / Federated Rules API
type: proposal
menu: proposals
status: accepted
owner: s-urbaniak,bwplotka
---

### Related Tickets

> https://github.com/thanos-io/thanos/pull/2200

Reviewer 3: I would argue that this is not related, it is your actual implementation since you started doing this backwards. This should probably have https://github.com/thanos-io/thanos/issues/1375.

Reviewer 2: That's true!

Author: Agreed I'll swap the referenced #2200 PR with https://github.com/thanos-io/thanos/issues/1375, although the latter is about federating targes but this proposal is about federating rules.

## Summary

Thanos allows a global query view for the Prometheus series. This means discovering, connecting various (often remote) "leafs" components and aggregating series data from them.

Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.

## Motivation

Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.

Pitfalls of current solutions:

> * It’s tedious to manually visit each leaf and we are lazy.

Reviewer 2: xD

    * This also means an up-to-date view (e.g statuses)
* It can be tied to Thanos.
* Simple to run if you already run Thanos.
    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
    * Potentially use the same connections, discovered targets by Querier.

## Verification

* Unit tests which would fire up a dummy RulesAPI and check different scenarios.
* Ad-hoc testing.

## Proposal

We propose to add following proto service and propagate all using existing components that has rules, so:
* Sidecar
* Querier (for federation)
* Ruler

The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.

```proto
service Rules {
  /// Rules has info for all rules.
  rpc Rules(RulesRequest) returns (stream RulesResponse);
}
```

Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.

> Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result.

Reviewer 3: How will this deduplication happen?

Reviewer 2: Good point, let's elaborate.

Author: Suggestion:

- Generally the deduplication logic is less complex than with time series, specifically:
- Deduplication happens first the rule group level. The identifier is the group name.
- Then, per group name deduplication happens on the rule level, where:
1. the rule type (recording rule vs. alerting rule)
2. the rule name
3. the rule label names

are being used as the deduplication identifer. Disjunct entries are simply merged by adding them to the result set.

Author: Added the above in the proposal text. Once we reach consensus and details I will amend.

Reviewer 3: Hmm, what if all of those things are the same but then the expression is a bit different in some way? That's still a valid setup i.e. nothing forbids you from doing that just like two different Prometheus replicas can "see" completely different metric values on the same target even though that is highly unlikely. Thus, we should probably take the expression itself also into consideration. What do you think? :)

Reviewer 3: Also, this could typically be caught by the label names but AFAICT external labels don't apply to what you get via `/api/v1/rules` - at least on my Prometheus nodes that does not happen.

Author: > Hmm, what if all of those things are the same but then the expression is a bit different in some way?

That is indeed a good catch :+1: I will amend the deduplication to consider the expression as an additional key.

Bear with me, this is my first contribution here, and it is already deep down.

Thanos query presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.

## Alternatives

* Cortex contains a sharded Ruler. Assigning rules to shards is done via Consul, though a gossip implementation is under development. Shards do not communicate with other shards. Rules come from a store (e.g. a Postgres database).

## Work Plan

* Implement a new flag `--rule` in Thanos Query which registers RulesAPI endpoints.
* Implement RulesAPI backends in sidecar, query, rule.
* Other StoreAPI backends implement a Noop RulesAPI backend, error'ing on RulesAPI invocations.

> * Feature branch: https://github.com/thanos-io/thanos/pull/2200

Reviewer 3: It's not a plan item if you had already started it, right? So I think this should be removed as well.

Reviewer 2: Still it was planned, so let's link it somehow, no? (:

Author: agreed, we should keep the link to the WIP PR somewhere :thinking:

Reviewer 3: Okay :+1: at least it's not at the beginning :smile:

Pitfalls of current solutions:
* It’s tedious to manually visit each leaf and we are lazy.
* Visiting leaf instances (be it Thanos Ruler or dedicated Prometheus endpoints) have to be implemented manually.

## Goals

* Present other Prometheus/Thanos Resources in a global view.
    * This also means an up-to-date view (e.g statuses)
* Simple to run if you already run Thanos.
    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
    * Potentially use the same connections, discovered targets by Querier.

## Verification

* Unit tests which would fire up a dummy Rules API and check different scenarios.
* Ad-hoc testing.

## Proposal

We propose to add following proto service and propagate all using existing components that has rules, so:
* Sidecar
* Querier (for federation)
* Ruler

> The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.

Reviewer 3: Small grammar nit:

```suggestion
The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alerting rules.
```

Author: indeed, fixed :+1:

```proto
service Rules {
  /// Rules has info for all rules.
  rpc Rules(RulesRequest) returns (stream RulesResponse);
}
```

> Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.

Reviewer 3: Some more grammar nits:

```suggestion
Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly, Thanos Ruler constructs the response based in its local state.
```

Author: good catched, fixed

> Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:

Reviewer 3: I wonder if it's really worth adding yet another flag. Because we know the StoreAPI types already perhaps it would be cleaner to use that information? :thinking: What does everyone else think about this idea?

Reviewer 2: I think I vote for another flag. Yes, it's +2 flags (probably we need `-strict` as well) but really it's just a flag - it has it's --help and it's readable. I think we need separation here. What if you want to use some server as Store API WITHOUT using it as RuleAPI? What if the opposite? I think different API needs different flag. (: 

Happy to hear other opinions.

Reviewer 3: In what use-cases you could want only to retrieve one or the other? :P Can't think of any off the top of my head. Another option would be to rename `--store` & friends to something like `--node` and then use gRPC reflection to know what services are exposed to us by the specified servers (https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#enable-server-reflection).

Nonetheless, we can go with the very explicit option that you are proposing here. I'm just afraid of the configuration complexity & verbosity. I think on our Slack we've talked a bit how we could implement another API for something else but I couldn't find the discussion now. So, we would get another pair of parameters in such a case? If not now then in the near future I think these should definitely be joined together.

Reviewer 2: > -store & friends to something like --node and then use gRPC reflection to know what services are exposed to us by the specified servers (https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#enable-server-reflection).

I was thinking initially that it's a good idea, but IMO it's not. It's better, especially for initial iteration to allow user for explicit configuration - always simpler. WDYT? What's the harm here?

Reviewer 2: yes the other one would be `TargetsAPI`

Reviewer 2: It is fair to talk about this now.. but I think explict `--target` would be needed as well

---
title: Global / Federated Rules API
type: proposal
menu: proposals
status: accepted
owner: s-urbaniak,bwplotka
---

### Related Tickets

https://github.com/thanos-io/thanos/issues/1375

## Summary

Thanos allows a global query view for the Prometheus series. This means discovering, connecting various (often remote) "leafs" components and aggregating series data from them.

Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.

## Motivation

> Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.

Reviewer 3: Seems like this wasn't solved. :smile:

> Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.

Author: oh gosh, you're right.

Pitfalls of current solutions:
* It’s tedious to manually visit each leaf and we are lazy.

> * Visiting leaf instances (be it ThanosRuler or dedicated Prometheus endpoints) have to be implemented manually.

Reviewer: ThanosRuler -> Thanos Ruler

## Goals

* Present other Prometheus/Thanos Resources in a global view.
    * This also means an up-to-date view (e.g statuses)

> * It can be tied to Thanos.

Reviewer: Not sure what do you mean here?

Reviewer 2: I am not sure as well ;p 

I think it can be "opinionated" and tailored for Thanos needs? I agree with @povilasv that it sounds weird as it is Thanos proposal after all...

Reviewer 2: Let's remove maybe

Author: agreed, removed

* Simple to run if you already run Thanos.
    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
    * Potentially use the same connections, discovered targets by Querier.

## Verification

> * Unit tests which would fire up a dummy RulesAPI and check different scenarios.

Reviewer: RulesAPI -> Rules API

* Ad-hoc testing.

## Proposal

We propose to add following proto service and propagate all using existing components that has rules, so:
* Sidecar
* Querier (for federation)
* Ruler

The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.

```proto
service Rules {
  /// Rules has info for all rules.

>   rpc Rules(RulesRequest) returns (stream RulesResponse);

Reviewer: Why stream? Wouldn't it me simpler to do an array?

Author: The intent was here to be in sync with Store API and potentially be prepared for future enhancement to be potentially stream based.

I do agree though, that in the case of rules this could potentially be a premature fore-sight.

Reviewer 2: I see the point of @povilasv here, but I fully agree with @s-urbaniak here.

My arguments for streaming: 
* Readability and consistency wins here as we already have tools for streaming dedup, partial response, timeouts and iterators so we can reuse more and all the patterns.
* It's will use bit less resources (e.g const merge time), but it's true as @povilasv said that we are not talking about high load here.

Author: ok, resolving the conversation as it seems we have reached consensus :-)

}
```

Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.

Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:

* Deduplication happens first the rule group level. The identifier is the group name.
* Then, per group name deduplication happens on the rule level, where:

1. the rule type (recording rule vs. alerting rule)
2. the rule name

> 3. the rule label names

Reviewer: Don't we need manual flag, like `--rule.replica-label=replica` be also part of identifier?

What I mean is scenario:

- Team1 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff which fires on their k8s namespace (different expr field, same labels)
- Team2 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff  which fires on their k8s namespace (different expr field, same labels)

In this case dedup would show a single rule?

I think this case might be quite common with monitoring mixins?

Author: Good catch, I believe this is the same edge case as https://github.com/thanos-io/thanos/pull/2263#discussion_r396521746.

I would suggest to add the expression as part of the deduplication key.

As for `--rule.replica-label=replica`: Good idea, but a counter-suggestion: Since rules api is a subset of store api, couldn't we reuse the same replica label, as the underlying metrics would have the same replica label? (just trying to prevent additional configuration complexity).

Reviewer 2: Wonder if this is related: https://github.com/coreos/prometheus-operator/issues/2193

We need to decide if we want to reuse querier replica labels or add another flag. I think it's kind of complex to make it more difficult to configure replicas (another format). Why not just another flag? Let's discuss, maybe we can find some format that would work for us.

`--replica=alerts{replica}` etc? This is what you mean?

Author: Suggestion while hopefully retaining enough orthogonality:

1. We do introduce a suggested `--rule.replica-label=replica` cli argument.
2. If that argument is not given we fall back to `--query.replica-label`.

But I am also very fine with simply a completely independent cli arg.

Author: @povilasv question to clear out ambiguities: Do you think that by introducing `--rule.replica-label` cli arg does not necessitate to consider the expression be part of the deduplication logic?

Reviewer: I think we still need expression, example alerts I have from kubernetes-mixin that have same exact labels names etc except for expression part:

```
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate1h) > (14.40 * 0.01000)
      and
      sum(apiserver_request:burnrate5m) > (14.40 * 0.01000)
    for: 2m
    labels:
      severity: critical
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate6h) > (6.00 * 0.01000)
      and
      sum(apiserver_request:burnrate30m) > (6.00 * 0.01000)
    for: 15m
    labels:
      severity: critical
```

Author: Indeed, very good example :+1: Adding the expression as part of the deduplication chain.

Author: In addition the `for` field should also be considered.

Reviewer 2: Wow, I did not that is possible 0.o

Reviewer 2: > Don't we need manual flag, like --rule.replica-label=replica be also part of identifier?
>
> What I mean is scenario:
>
> Team1 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff which fires on their k8s namespace (different expr field, same labels)
> Team2 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff which fires on their k8s namespace (different expr field, same labels)
> In this case dedup would show a single rule?
> 
> I think this case might be quite common with monitoring mixins?

I don't think we need separate replica labels. Let's not complicate things as,as you can see it can escalate quickly: https://github.com/thanos-io/thanos/pull/2263#discussion_r406109403

In the case you describes @povilasv Thanos Rule has to have unique labels as Prometheus external labels, so this is solved.

are being used as the deduplication identifer. Disjunct entries are simply merged by adding them to the result set.

Thanos query presents the result of the fan-out on a `/api/v1/rules` endpoint which is compatible with the Prometheus' rules API endpoint.

## Alternatives

* Cortex contains a sharded Ruler. Assigning rules to shards is done via Consul, though a gossip implementation is under development. Shards do not communicate with other shards. Rules come from a store (e.g. a Postgres database).

## Work Plan

* Implement a new flag `--rule` in Thanos Query which registers RulesAPI endpoints.
* Implement RulesAPI backends in sidecar, query, rule.

> * Other StoreAPI backends implement a Noop RulesAPI backend, error'ing on RulesAPI invocations.

Reviewer: Why do we need a Noop Rules API? Can we just not implement it?

Author: I don't think so, because other Thanos components do not have any means of providing Rules API.

Reviewer 2: Actually that's a good catch @povilasv  - Why we need to implement anything for those? Just do nothing for those (:

Author: I missed the "not" :man_facepalming: :-D

Reviewer 3: :smile: seems like this was solved so marking as "Resolved"

Reviewer: This sentence is still wrong :/
There is no need to do `Other StoreAPI backends implement a Noop RulesAPI backend, error'ing on RulesAPI invocations.` 

can we just remove it?

Author: ah yes, indeed, removed.
